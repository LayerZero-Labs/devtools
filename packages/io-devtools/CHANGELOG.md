# @layerzerolabs/io-devtools

## 0.1.10

### Patch Changes

- 8b32a79: Use less strict version requirements for layerzero peer dependencies

## 0.1.9

### Patch Changes

- 938ac3d: Improve error reporting on UIntBigIntSchema

## 0.1.8

### Patch Changes

- 9af95ca: Add createWithAsyncLogger logging utility for async functions

## 0.1.7

### Patch Changes

- 8a4addf: Replace cli-table3 with table

## 0.1.6

### Patch Changes

- 2a36974: Add the ability to define configs as functions

## 0.1.5

### Patch Changes

- 54cf16e: Silence bigint-buffer warning

## 0.1.4

### Patch Changes

- 9e78abe: Update hardhat-deploy

## 0.1.3

### Patch Changes

- e766c86: Adding in Default Colors to print outs

## 0.1.2

### Patch Changes

- 253c79e: Rename description to hint for select prompts
- 5c58d69: Update @layerzerolabs dependencies to 2.1.3

## 0.1.1

### Patch Changes

- 40e8dba: Remove CommaSeparatedValuesSchema in favor of splitCommaSeparated; Move LogLevelSchema; Export isLogLevel

## 0.1.0

### Minor Changes

- 120adf1: Make packages public

## 0.0.7

### Patch Changes

- 81d7eee: Add promptForText utility

## 0.0.6

### Patch Changes

- 2b36ed9: Updating Table for Get OApp Config Task
- 2b36ed9: Rafactor some of the printers in io-devtools

## 0.0.5

### Patch Changes

- ed450c4: Add promptToSelectOne and promptToSelectMultiple utilities

## 0.0.4

### Patch Changes

- af8cc25: Update dependencies

## 0.0.3

### Patch Changes

- 5077bf1: Unmount the component when using printTable

## 0.0.2

### Patch Changes

- e0be8b7: Reexport Logger type
- 70646b4: Add missing dependencies

## 0.0.1

### Patch Changes

- b74afbe: Initial 0.0.1 version
