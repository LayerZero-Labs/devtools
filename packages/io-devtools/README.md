<p align="center">
  <a href="https://layerzero.network">
    <img alt="LayerZero" style="max-width: 500px" src="https://d3a2dpnnrypp5h.cloudfront.net/bridge-app/lz.png"/>
  </a>
</p>

<h1 align="center">@layerzerolabs/io-utils</h1>

<!-- The badges section -->
<p align="center">
  <!-- Shields.io NPM published package version -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM Version" src="https://img.shields.io/npm/v/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io NPM downloads -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="Downloads" src="https://img.shields.io/npm/dm/@layerzerolabs/io-utils"/></a>
  <!-- Shields.io license badge -->
  <a href="https://www.npmjs.com/package/@layerzerolabs/io-utils"><img alt="NPM License" src="https://img.shields.io/npm/l/@layerzerolabs/io-utils"/></a>
</p>

## Installation

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
pnpm install @layerzerolabs/io-utils
```

```bash
npm install @layerzerolabs/io-utils
```

## API Documentation

### Filesystem utilities

#### isDirectory(path)

Returns `true` if specified filesystem `path` points to a directory, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isFile(path)

Returns `true` if specified filesystem `path` points to a file, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

#### isReadable(path)

Returns `true` if specified filesystem `path` can be read by the current user, `false` otherwise. Does not throw if `path` does not exist on the filesystem, instead returns `false`

### Standard input/output utilities

#### promptToContinue([message, defaultValue])

Asks the user whether they want to continue and reads the input from the CLI standard input. By default the question displayed is `Do you want to continue?` and the default response is `yes`

```typescript
const goahead = await promptToContinue();

// To ask a different question
const goahead = await promptToContinue("Are you sure?");

// To default the response to false, good for important and unsafe decisions
const goahead = await promptToContinue("Are you sure?", false);
```
