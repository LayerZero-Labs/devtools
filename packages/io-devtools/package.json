{
  "name": "@layerzerolabs/io-devtools",
  "version": "0.1.10",
  "description": "Utilities for working with I/O in LayerZero devtools",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/LayerZero-Labs/devtools.git",
    "directory": "packages/io-devtools"
  },
  "license": "MIT",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": {
      "types": "./dist/*/index.d.ts",
      "require": "./dist/*/index.js",
      "import": "./dist/*/index.mjs"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "swag"
  ],
  "scripts": {
    "prebuild": "tsc -noEmit",
    "build": "$npm_execpath tsup",
    "clean": "rm -rf dist",
    "dev": "$npm_execpath tsup --watch",
    "lint": "$npm_execpath eslint '**/*.{js,ts,json}'",
    "lint:fix": "eslint --fix '**/*.{js,ts,json}'",
    "test": "jest --ci"
  },
  "dependencies": {
    "chalk": "^4.1.2",
    "logform": "^2.6.0",
    "prompts": "^2.4.2",
    "table": "~6.8.2",
    "winston": "^3.11.0"
  },
  "devDependencies": {
    "@swc/core": "^1.4.0",
    "@swc/jest": "^0.2.36",
    "@types/ink-gradient": "^2.0.4",
    "@types/jest": "^29.5.12",
    "@types/prompts": "^2.4.9",
    "@types/react": "^17.0.75",
    "fast-check": "^3.15.1",
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "jest": "^29.7.0",
    "jest-extended": "^4.0.2",
    "react": "^17.0.2",
    "ts-node": "^10.9.2",
    "tslib": "~2.6.2",
    "tsup": "~8.0.1",
    "typescript": "^5.3.3",
    "zod": "^3.22.4"
  },
  "peerDependencies": {
    "ink": "^3.2.0",
    "ink-gradient": "^2.0.0",
    "ink-table": "^3.1.0",
    "react": "^17.0.2",
    "yoga-layout-prebuilt": "^1.9.6",
    "zod": "^3.22.4"
  },
  "peerDependenciesMeta": {
    "ink": {
      "optional": true
    },
    "ink-gradient": {
      "optional": true
    },
    "ink-table": {
      "optional": true
    },
    "react": {
      "optional": true
    },
    "yoga-layout-prebuilt": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public"
  }
}
