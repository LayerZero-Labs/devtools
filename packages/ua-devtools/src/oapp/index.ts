export * from './check'
export * from './config'
export * from './schema'
export * from './types'
