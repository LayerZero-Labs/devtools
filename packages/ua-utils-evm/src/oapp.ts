import { Contract } from "@ethersproject/contracts"
import { GetPropertyValue, createProperty } from "@layerzerolabs/ua-utils"
import { EndpointId } from "@layerzerolabs/lz-definitions"

type SetPeerConfigurableContext = [oapp: Contract, endpointId: EndpointId]

type SetPeerConfigurableValue = string

export const createSetPeerProperty = (desired: GetPropertyValue<SetPeerConfigurableContext, SetPeerConfigurableValue>) =>
    createProperty<SetPeerConfigurableContext, SetPeerConfigurableValue>({
        desired,
        get: (oapp, endpointId) => oapp.peers(endpointId),
        set: (oapp, endpointId, peer) => oapp.setPeer(endpointId, peer),
    })
