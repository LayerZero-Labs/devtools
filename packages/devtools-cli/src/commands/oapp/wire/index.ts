import { Command } from 'commander'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { createLogger, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import {
    type WithLogLevelOption,
    type WithSetupOption,
    WithTsConfigOption,
    createLogLevelOption,
    createSetupFileOption,
    createTsConfigFileOption,
} from '@/commands/options'
import { setupTypescript } from '@/setup'

interface Args extends WithLogLevelOption, WithSetupOption, WithTsConfigOption {}

export const wire = new Command('wire')
    .addOption(createSetupFileOption())
    .addOption(createTsConfigFileOption())
    .addOption(createLogLevelOption())
    .action(async ({ setup, logLevel, tsConfig: tsConfigPath }: Args) => {
        printLogo()

        // We'll set the global logging level to get as much info as needed
        setDefaultLogLevel(logLevel)

        // We'll setup TypeScript support so that we can dynamically load TypeScript config files
        setupTypescript(tsConfigPath)

        const logger = createLogger(logLevel)

        logger.debug(`Loading setup from ${setup}`)

        logger.warn(
            `This command is just a placeholder. Please use @layerzerolabs/toolbox-hardhat package for the time being.`
        )
    })
