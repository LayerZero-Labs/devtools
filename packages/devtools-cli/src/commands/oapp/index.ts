import { Command } from 'commander'
import { wire } from './wire'

export const oapp = new Command('oapp').description('OApp configuration commands').addCommand(wire)
