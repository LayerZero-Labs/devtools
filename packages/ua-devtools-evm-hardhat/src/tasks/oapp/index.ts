import './wire'
import './peers.get'
import './config.get.default'
import './config.get.executor'
import './config.get'
import './enforced.opts.get'
import './config.init'
import './subtask.config.load'
