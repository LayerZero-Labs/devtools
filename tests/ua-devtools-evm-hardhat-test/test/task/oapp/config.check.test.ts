import hre from 'hardhat'
import { resolve } from 'path'
import { isFile } from '@layerzerolabs/io-devtools'
import { deployOAppFixture } from '../../__utils__/oapp'
import { TASK_LZ_OAPP_CONFIG_CHECK, TASK_LZ_OAPP_WIRE } from '@layerzerolabs/ua-devtools-evm-hardhat'

describe(`task ${TASK_LZ_OAPP_CONFIG_CHECK}`, () => {
    const CONFIGS_BASE_DIR = resolve(__dirname, '__data__', 'configs')
    const configPathFixture = (fileName: string): string => {
        const path = resolve(CONFIGS_BASE_DIR, fileName)
        expect(isFile(path)).toBeTruthy()
        return path
    }

    beforeEach(async () => {
        await deployOAppFixture()
    })

    it('should show no chains are connected', async () => {
        const oappConfig = configPathFixture('valid.config.connected.js')
        const result = await hre.run(TASK_LZ_OAPP_CONFIG_CHECK, { oappConfig })
        const expectPoint = { eid: expect.any(Number), address: expect.any(String) }
        const expectVector = { from: expectPoint, to: expectPoint }
        expect(result).toEqual([
            { vector: expectVector, hasPeer: false },
            { vector: expectVector, hasPeer: false },
        ])
    })

    it('should show both chains are connected after running wire', async () => {
        const oappConfig = configPathFixture('valid.config.connected.js')
        await hre.run(TASK_LZ_OAPP_WIRE, { oappConfig, ci: true })
        const result = await hre.run(TASK_LZ_OAPP_CONFIG_CHECK, { oappConfig })
        const expectPoint = { eid: expect.any(Number), address: expect.any(String) }
        const expectVector = { from: expectPoint, to: expectPoint }
        expect(result).toEqual([
            { vector: expectVector, hasPeer: true },
            { vector: expectVector, hasPeer: true },
        ])
    })
})
