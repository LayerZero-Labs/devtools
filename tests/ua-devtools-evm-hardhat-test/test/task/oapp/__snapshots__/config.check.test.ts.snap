// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task lz:oapp:config:check should show all chains are connected after running wire with three networks 1`] = `
[
  "┌───────────┬─────────┬───────────┬───────┐
│ from → to │ britney │ vengaboys │ tango │
├───────────┼─────────┼───────────┼───────┤
│ britney   │ ∅       │ ✓         │ ✓     │
├───────────┼─────────┼───────────┼───────┤
│ vengaboys │ ✓       │ ∅         │ ✓     │
├───────────┼─────────┼───────────┼───────┤
│ tango     │ ✓       │ ✓         │ ∅     │
└───────────┴─────────┴───────────┴───────┘",
  "
	✓ - Connected
",
  "	⤫ - Not Connected
",
  "	∅ - Ignored",
]
`;

exports[`task lz:oapp:config:check should show all chains are connected after running wire with two networks 1`] = `
[
  "┌───────────┬─────────┬───────────┐
│ from → to │ britney │ vengaboys │
├───────────┼─────────┼───────────┤
│ britney   │ ∅       │ ✓         │
├───────────┼─────────┼───────────┤
│ vengaboys │ ✓       │ ∅         │
└───────────┴─────────┴───────────┘",
  "
	✓ - Connected
",
  "	⤫ - Not Connected
",
  "	∅ - Ignored",
]
`;

exports[`task lz:oapp:config:check should show all chains are connected expect one after running wire with three networks 1`] = `
[
  "┌───────────┬─────────┬───────────┬───────┐
│ from → to │ britney │ vengaboys │ tango │
├───────────┼─────────┼───────────┼───────┤
│ britney   │ ∅       │ ✓         │ ✓     │
├───────────┼─────────┼───────────┼───────┤
│ vengaboys │ ✓       │ ∅         │ ✓     │
├───────────┼─────────┼───────────┼───────┤
│ tango     │ ∅       │ ✓         │ ∅     │
└───────────┴─────────┴───────────┴───────┘",
  "
	✓ - Connected
",
  "	⤫ - Not Connected
",
  "	∅ - Ignored",
]
`;

exports[`task lz:oapp:config:check should show all chains are not connected expect one after running wire with three networks 1`] = `
[
  "┌───────────┬─────────┬───────────┬───────┐
│ from → to │ britney │ vengaboys │ tango │
├───────────┼─────────┼───────────┼───────┤
│ britney   │ ∅       │ ∅         │ ∅     │
├───────────┼─────────┼───────────┼───────┤
│ vengaboys │ ✓       │ ∅         │ ∅     │
├───────────┼─────────┼───────────┼───────┤
│ tango     │ ∅       │ ∅         │ ∅     │
└───────────┴─────────┴───────────┴───────┘",
  "
	✓ - Connected
",
  "	⤫ - Not Connected
",
  "	∅ - Ignored",
]
`;

exports[`task lz:oapp:config:check should show no chains are connected with two networks 1`] = `
[
  "┌───────────┬─────────┬───────────┐
│ from → to │ britney │ vengaboys │
├───────────┼─────────┼───────────┤
│ britney   │ ∅       │ ⤫         │
├───────────┼─────────┼───────────┤
│ vengaboys │ ⤫       │ ∅         │
└───────────┴─────────┴───────────┘",
  "
	✓ - Connected
",
  "	⤫ - Not Connected
",
  "	∅ - Ignored",
]
`;
